From d6a3b53e02b84620610bb338aed42de0223c0d6e Mon Sep 17 00:00:00 2001
From: Evan Wang <xswang@marvell.com>
Date: Thu, 17 Aug 2017 19:11:37 +0800
Subject: [PATCH 1197/1345] phy: comphy: CP110: add PIPE selector clear
 support

commit  128473ae11d84032d0c1bb350ab528871a2db008 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

Till now both PCIe and USB are supported in CP110 COMPHY
driver, and it is time to add the PIPE selector
clear function to clear the selector when COMPHY
powered off, which remove the depedency of uboot.

Change-Id: Idb1bee67df061356d12d7c65d15946bdd7b85aba
Signed-off-by: Evan Wang <xswang@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/43163
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/phy/phy-comphy-cp110.c |   21 ++++++++++++++++++++-
 1 file changed, 20 insertions(+), 1 deletion(-)

diff --git a/drivers/phy/phy-comphy-cp110.c b/drivers/phy/phy-comphy-cp110.c
index daf68bd..bd219cd 100644
--- a/drivers/phy/phy-comphy-cp110.c
+++ b/drivers/phy/phy-comphy-cp110.c
@@ -123,6 +123,24 @@ static void mvebu_cp110_comphy_set_phy_selector(struct mvebu_comphy_priv *priv,
 
 }
 
+/* Clear PIPE selector - avoid collision with prior u-boot configuration */
+void mvebu_cp110_comphy_clr_pipe_selector(struct mvebu_comphy_priv *priv,
+					  struct mvebu_comphy *comphy)
+{
+	u32 reg, mask, field;
+	u32 comphy_offset = COMMON_SELECTOR_COMPHYN_FIELD_WIDTH * comphy->index;
+
+	mask = COMMON_SELECTOR_COMPHY_MASK << comphy_offset;
+	reg = readl(priv->comphy_regs + COMMON_SELECTOR_PIPE_REG_OFFSET);
+	field = reg & mask;
+
+	if (field) {
+		reg &= ~mask;
+		writel(reg,
+		       priv->comphy_regs + COMMON_SELECTOR_PIPE_REG_OFFSET);
+	}
+}
+
 /* PIPE selector configures for PCIe, USB 3.0 Host, and USB 3.0 Device mode */
 void mvebu_cp110_comphy_set_pipe_selector(struct mvebu_comphy_priv *priv,
 					  struct mvebu_comphy *comphy)
@@ -1638,8 +1656,9 @@ static int mvebu_cp110_comphy_power_off(struct phy *phy)
 		break;
 	}
 
-	/* Clear comphy selector, can't rely on u-boot */
+	/* Clear comphy PHY and PIPE selector, can't rely on u-boot */
 	mvebu_cp110_comphy_clr_phy_selector(priv, comphy);
+	mvebu_cp110_comphy_clr_pipe_selector(priv, comphy);
 
 	spin_unlock(&priv->lock);
 
-- 
1.7.9.5

